src/secp256k1.c

```diff
@@ -462,34 +485,36 @@ int secp256k1_ecdsa_sign(const secp256k1_context* ctx, secp256k1_ecdsa_signature
 
     secp256k1_scalar_set_b32(&sec, seckey, &overflow);
     /* Fail if the secret key is invalid. */
-    if (!overflow && !secp256k1_scalar_is_zero(&sec)) {
-        unsigned char nonce32[32];
-        unsigned int count = 0;
-        secp256k1_scalar_set_b32(&msg, msg32, NULL);
-        while (1) {
-            ret = noncefp(nonce32, msg32, seckey, NULL, (void*)noncedata, count);
-            if (!ret) {
+    overflow |= secp256k1_scalar_is_zero(&sec);
+    secp256k1_scalar_cmov(&sec, &secp256k1_scalar_one, overflow);
+    secp256k1_scalar_set_b32(&msg, msg32, NULL);
+    while (1) {
+        int koverflow;
+        ret = noncefp(nonce32, msg32, seckey, NULL, (void*)noncedata, count);
+        if (!ret) {
+            break;
+        }
+        secp256k1_scalar_set_b32(&non, nonce32, &koverflow);
+        koverflow |= secp256k1_scalar_is_zero(&non);
+        /* The nonce is still secret here, but it overflowing or being zero is is less likely than 1:2^255. */
+        declassify(ctx, &koverflow, sizeof(koverflow));
+        if (!koverflow) {
```

Comment:
Letting it overflow would bias the nonce, which is *extremely* bad for security.  The order of secp256k1 is fortunately such that the bias would not actually be a problem, though this isn't the case for many other curves.  Personally, if I were auditing an someone elses crypto code and saw it producing biased nonces like that I would instant fail the audit and not bother looking at anything else because the software was clearly written people by people who didn't notice or didn't heed the pile of dead bodies of their forebearers that thought leaving nonces a little biased was okay.

There isn't much point in cmoving there because the loop must run again that case, making the function take a non-constant time. So the cmov would just be constant time theatre with no improvement. It would _still_ have to be declassified too,  since constant time analysis would notice the loop being conditioned on it-- so the cmov would be just pure theater, not actually improve the constant timeness and just adding code.

Reply:
> Personally, if I were auditing an someone elses crypto code and saw it producing biased nonces like that I would instant fail the audit and not bother looking at anything else because the software was clearly written people by people who didn't notice or didn't heed the pile of dead bodies of their forebearers that thought leaving nonces a little biased was okay.

We ignore the overflow in the BIP 340 draft by the way. :X

Reply:
@real-or-random I think the situation is a little different in BIP340, because it's not using a generic retrying PRG-based approach, but a specified scheme (and curve) specific algorithm, with a rationale explaining that this approach is only safe because the order is so close to 2^256.

If we'd decide that's worrisome because it may end up being reusing for another curve, I'd prefer the EdDSA approach (hash to twice the width, and then modulo reduce it) over an iterative approach.